src/xdai-bots/x-kleros-liquid.js

```diff
@@ -0,0 +1,178 @@
+
+          const notTieAndNoOneCoherent = voteCounters.map(
+            v =>
+              !voteCounters[voteCounters.length - 1].tied &&
+              v.counts[voteCounters[voteCounters.length - 1].winningChoice] ===
+                '0'
+          )
+          if (
+            !dispute.ruled ||
+            dispute2.votesLengths.some(
+              (l, i) =>
+                Number(notTieAndNoOneCoherent[i] ? l : l * 2) !==
+                Number(dispute2.repartitionsInEachRound[i])
+            )
+          )
+            // The dispute is not finalized, try to call all of its callbacks.
+            batchedSend(
+              [
+                // We check if there are still pending draws because if there aren't any and the dispute is still in the evidence period,
+                // then the transaction would still succeed and we don't want to waste gas in those cases.
+                dispute2.votesLengths[dispute2.votesLengths.length - 1] >
+                  dispute.drawsInRound && {
+                  args: [disputeID, 15],
+                  method: klerosLiquid.methods.drawJurors,
+                  to: klerosLiquid.options.address
+                },
+                ...dispute2.votesLengths.map(
+                  // eslint-disable-next-line no-loop-func
+                  (l, i) =>
+                    Number(notTieAndNoOneCoherent[i] ? l : l * 2) !==
+                      Number(dispute2.repartitionsInEachRound[i]) && {
+                      args: [disputeID, i, 15],
+                      method: klerosLiquid.methods.execute,
+                      to: klerosLiquid.options.address
+                    }
+                ),
+                {
+                  args: [disputeID],
+                  method: klerosLiquid.methods.executeRuling,
+                  to: klerosLiquid.options.address
+                },
+                {
+                  args: [disputeID],
+                  method: klerosLiquid.methods.passPeriod,
+                  to: klerosLiquid.options.address
+                }
+              ].filter(t => t)
+            )
+          else executedDisputeIDs[disputeID] = true // The dispute is finalized, cache it.
+        }
+        disputeID++
+      }
+    } catch (_) {} // Reached the end of the disputes list.
+
+    // Try to pass the phase.
+    let readyForNextPhase = false
+    const phase = await klerosLiquid.methods.phase().call()
+    const lastPhaseChange = await klerosLiquid.methods.lastPhaseChange().call()
+    const disputesWithoutJurors = await klerosLiquid.methods
+      .disputesWithoutJurors()
+      .call()
+    if (phase == PhaseEnum.staking) {
+      const minStakingTime = await klerosLiquid.methods.minStakingTime().call()
+      if (
+        Date.now() - lastPhaseChange * 1000 >= minStakingTime * 1000 &&
+        disputesWithoutJurors > 0
+      ) {
+        readyForNextPhase = true
+      }
+    } else if (phase == PhaseEnum.generating) {
+      readyForNextPhase = true
```

Comment:
This needs some extra attention.
On mainnet this makes sense because we need only one new block to generate the random number. On xDai we need between 76 and 152 blocks ( = 5 seconds per block * 152 blocks = 13 minutes). Not a big deal to have a few tx reverting on xDai, but the additional problem is that is not allowed to retrieve a number during the reveal phase.

Reply:
I'm not sure what's the best way to solve this. Maybe query the RandomAuRa contract and check if it's in the commitPhase(). If not and current block is greater than RNBlock, stay there looping and querying until it's time.

Reply:
Or simply try sending the transaction over and over again and don't worry if it reverts. I would recommend that for this special case we use a greater frequency (1 min or so).

Reply:
> If not and current block is greater than RNBlock, stay there looping and querying until it's time.

The problem I see with this is that the action callback bot runs every 5 minutes. This would create overlapping jobs, which would eventually exhaust the VM, killing all other bots as a side-effect.
Maybe it's better to just let the txn fail or to check if it can call it already.